histogram/src/lib.rs

```diff
@@ -1,206 +1,155 @@
-
-        // this histogram config doesn't have much resolution, which results in
-        // the upper percentiles falling into buckets that are rather wide
-        assert_eq!(
-            &percentiles,
-            &[
-                (8, 11),
-                (96, 127),
-                (256, 383),
-                (512, 767),
-                (768, 1023),
-                (768, 1023),
-                (768, 1023)
-            ]
-        );
+        Some(Bucket {
+            count: self.get_count(index),
+            lower: self.config().index_to_lower_bound(index),
+            upper: self.config().index_to_upper_bound(index),
+        })
     }
 
-    #[test]
-    fn percentiles_2() {
-        let histogram = Histogram::new(0, 5, 10).unwrap();
-
-        for v in 1..1024 {
-            assert!(histogram.increment(v, 1).is_ok());
-            assert_eq!(histogram.percentile(0.0).map(|b| b.high()), Ok(1));
+    fn percentiles(&self, percentiles: &[f64]) -> Result<Vec<(f64, Bucket)>, Error> {
+        // get the total count across all buckets as a u64
+        let total: u128 = self.total_count();
 
-            assert!(histogram.percentile(100.0).map(|b| b.high()).unwrap_or(0) >= v);
-            assert!(histogram.percentile(100.0).map(|b| b.low()).unwrap_or(0) <= v);
+        // if the histogram is empty, then we should return an error
+        if total == 0_u128 {
+            // TODO(brian): this should return an error =)
+            return Err(Error::Empty);
         }
 
-        let percentiles: Vec<(u64, u64)> = histogram
-            .percentiles(&[1.0, 10.0, 25.0, 50.0, 75.0, 90.0, 99.0])
-            .unwrap()
-            .iter()
-            .map(|p| (p.bucket().low(), p.bucket().high()))
-            .collect();
-
-        // this histogram config has enough resolution to keep the error lower
-        assert_eq!(
-            &percentiles,
-            &[
-                (11, 11),
-                (100, 103),
-                (256, 271),
-                (512, 543),
-                (768, 799),
-                (896, 927),
-                (992, 1023)
-            ]
-        );
-    }
-
-    #[test]
-    fn percentiles_3() {
-        let histogram = Histogram::builder().build().unwrap();
-        histogram.increment(1, 1).unwrap();
-        histogram.increment(10000000, 1).unwrap();
-
-        let percentiles = histogram.percentiles(&[25.0, 75.0]).unwrap();
-
-        assert_eq!(histogram.percentile(25.0).map(|b| b.high()), Ok(1));
-        assert_eq!(histogram.percentile(75.0).map(|b| b.high()), Ok(10010623));
-
-        assert_eq!(percentiles.get(0).map(|b| b.bucket().high()), Some(1));
-        assert_eq!(
-            percentiles.get(1).map(|b| b.bucket().high()),
-            Some(10010623)
-        );
-    }
+        // sort the requested percentiles so we can find them in a single pass
+        let mut percentiles = percentiles.to_vec();
+        percentiles.sort_by(|a, b| a.partial_cmp(b).unwrap());
```

Comment:
might as well convert percentiles to counts at this step and remove the ` / 100 * count` step for each percentile in the outer loop.

Reply:
We still need the actual percentiles to form the result vec. However, we can precompute the counts. Please see latest version.